cmd/contour/shutdownmanager.go

```diff
@@ -0,0 +1,132 @@
+// Copyright © 2020 VMware
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+	"time"
+
+	"github.com/projectcontour/contour/internal/metrics"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/projectcontour/contour/internal/workgroup"
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+// handler for /healthz
+func (s *shutdownmanagerContext) healthzHandler(w http.ResponseWriter, r *http.Request) {
+	w.WriteHeader(http.StatusOK)
+	_, err := w.Write([]byte("OK"))
+	if err != nil {
+		s.Error(err)
+	}
+}
+
+// handles /shutdown
+func (s *shutdownmanagerContext) shutdownHandler(w http.ResponseWriter, r *http.Request) {
+	prometheusURL := fmt.Sprintf("http://%s:%d%s", s.envoyHost, s.envoyPort, s.prometheusPath)
+	envoyAdminURL := fmt.Sprintf("http://%s:%d/healthcheck/fail", s.envoyHost, s.envoyPort)
+
+	// Send shutdown signal to Envoy to start draining connections
+	err := shutdownEnvoy(envoyAdminURL)
```

Comment:
My thought was to be optimistic and keep trying. Worst case if the stats endpoint can't be found, it will still wait for the total termination grace period giving the most amount of time for connections to drain.

Reply:
Yeh, I think that the stats behavior sense.

Should the shutdown retry too? If the shutdown transiently fails, then sampling the stats is futile. OTOH, if the shutdown can fail for permanent reasons (e.g. envoy is already shutting down), then we would get stuck here.